### DISCUSSION TITLE
Add TM1638 switches and relays to display build

### DISCUSSION THREAD

**Discussion by jsiddall** [arendst/Tasmota#21557] [Ideas]
Currently switches and relays on the TM1638 module work in the tasmota32 builds but not in the tasmota32-display builds. This is odd because the TM1638 is a display module so the display build should support all the module capabilities.

Can switches and relays on the TM1638 module also be added to the tasmota32-display builds?

This was originally opened as: https://github.com/arendst/Tasmota/issues/21554 but was closed and re-opened here

**Comment by Jason2866** [comment]
Tbh don't understand why. Just use the build where TM1638 is supported. 
Variant Display is far away from supporting all displays only the most used ones.
In general trying to avoid to have drivers duplicate in variants.
We see the TM1638 in first step as port expander ;-)

**Reply by jsiddall** [reply]
Yes, the TM1638 is a hybrid device: it has 8 switches, 8 LEDs (which Tasmota represents as 8 relays) and 8 digits of 7-segment display (using the TM1637 display driver).

Tasmota supports all 3 of these capabilities, but none of the current builds supports all 3.  The regular tasmota32 build (and the sensors build) supports the switches and the relays, but not the 7-segment display.  The display builds supports the 7-segment display but not the switches or relays.

I would like to use all 3 in one build and I think it makes sense for this to be the tasmota32-display build.  The request is just to _add_  the ability to use the switches and the relays that are included in the regular tasmota32 build to the tasmota32-display build.

**Comment by Jason2866** [comment]
@arendst What do you think? Since we gain some space in the display build we could add the not big switches and relay driver to the display variant (knowing that it is not fully correct "placed")

**Reply by arendst** [reply]
ok with me.

**Comment by jsiddall** [comment]
For reference, here is a picture of the board:

![board](https://tasmota.github.io/docs/_media/peripherals/TM1638-front.png)

Very inexpensive way of testing with switches, relays and a display

**Comment by Jason2866** [comment]
Added in PR https://github.com/arendst/Tasmota/pull/21617
